### PR TITLE
DART example

### PR DESCRIPTION
**Pull request by costashatz**
This PR adds an example of using the [DartIntegration](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1DartIntegration.html). The example consists of a KUKA Iiwa (14kg payload -- model got from [here](https://github.com/epfl-lasa/iiwa_ros)) manipulator equipped with a Robotiq 2-finger gripper (model got from [here](https://github.com/a-price/robotiq_arg85_description)) that the user can control to make some small towers of boxes. Here's an example of how a completed tower would look like:
image

The code also includes some basic operational space control for manipulators. In the blogpost that will come, I will explain also in detail the robotics aspects and the model-based controller.

This PR is almost finished; I think I only need to add a few lines of documentation and explanation on how to use it.. @mosra let me know what you think and sorry for being late..

### CODE REVIEW

File: src/dart/DartExample.cpp [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+        const double _pLinearGain = 500.;
+        const double _dLinearGain = 200.;
+        const double _pOrientationGain = 100.;
+        const double _dOrientationGain = 50.;
+        const double _dRegularization = 10.;
+        const double _pGain = 2.25;
+        const double _dGain = 5.;
+
+        /* Simple State Machine */
+        size_t state = 0; /* 0: go to home position, 1: go to active box */
```
Comment (mosra) [inline src/dart/DartExample.cpp]: Maybe an enum for this? Also, `_state`?

File: src/dart/DartExample.cpp [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+        .setViewport(GL::defaultFramebuffer.viewport().size());
+    /* DART has +Z-axis as up direction*/
+    _cameraObject->setTransformation(Magnum::Matrix4::lookAt(Vector3{0.f, 3.f, 1.5f}, Vector3{0.f, 0.f, 0.5f}, Vector3{0.f, 0.f, 1.f}));
+
+    /* DART: Load Skeletons/Robots */
+    DartLoader loader;
+    /* Add packages (needed for URDF loading) */
+    loader.addPackageDirectory("iiwa_description", std::string(DARTEXAMPLE_DIR) + "/urdf/");
+    loader.addPackageDirectory("robotiq_arg85_description", std::string(DARTEXAMPLE_DIR) + "/urdf/");
+    std::string filename = std::string(DARTEXAMPLE_DIR) + "/urdf/iiwa14_simple.urdf";
```
Comment (mosra) [inline src/dart/DartExample.cpp]: Could be `DARTEXAMPLE_DIR "/urdf/"` (the preprocessor concatenates the literals), but I think I'd rather see `Utility::Directory::join()` here :)

File: src/dart/DartExample.cpp [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+    UnsignedInt steps = std::ceil(dt / _world->getTimeStep());
+
+    /* Step DART simulation */
+    for (UnsignedInt i = 0; i < steps; i++) {
+        updateManipulator();
+        _dartWorld->step();
+    }
+
+    /* Update graphic meshes/materials and render */
+    updateGraphics();
```
Comment (mosra) [inline src/dart/DartExample.cpp]: I think the contents of `updateManipulator()` and `updateGraphics()` could be both directly here I think, no need to have them in a separate function (that's only called from a single place anyway). The current way makes it look like there's "almost nothing" done in the draw event, which is far from true :)

File: src/dart/DartExample.cpp [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+    at large and to the detriment of our heirs and successors. We intend this
+    dedication to be an overt act of relinquishment in perpetuity of all
+    present and future rights to this software under copyright law.
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+    IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+    CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExample.cpp]: Should be fixed now!

File: src/dart/DartExampleHelpers.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#ifndef MAGNUmatEXAMPLES_DART_EXAMPLE_HELPERS_H
+#define MAGNUmatEXAMPLES_DART_EXAMPLE_HELPERS_H
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExampleHelpers.h]: In the beginning, I had many more!! Removed the file!

File: src/dart/DartExample.cpp [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+    Debug{} << "Press 'r' to go above the red box (switches to control mode)";
+    Debug{} << "Press 'g' to go above the green box (switches to control mode)";
+    Debug{} << "Press 'b' to go above the blue box (switches to control mode)";
+    Debug{} << "Press 'c' to close the gripper";
+    Debug{} << "Press 'o' to open the gripper";
+    Debug{} << "Press 'd' to go further down (only in control mode)";
+    Debug{} << "Press 'u' to go further up (only in control mode)";
+    Debug{} << "Press 'z' or 'x' to move sideways (only in control mode)";
+    Debug{} << "Press 'h' to go to home position (switches to idle mode)";
+    Debug{} << "Press 'space' to reset the world";
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExample.cpp]: Created a doc file (+ image!)..

File: src/dart/DartExample.cpp [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+    _mesh->draw(*_shader);
+    if(_isSoftBody)
+        GL::Renderer::enable(GL::Renderer::Feature::FaceCulling);
+}
+
+
+ColoredObject& ColoredObject::setMesh(GL::Mesh* mesh){
+    _mesh = mesh;
+    return *this;
+}
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExample.cpp]: You're right!

File: src/dart/DartExample.cpp [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+    if(event.key() == KeyEvent::Key::Down)
+        _cameraObject->rotateX(Deg(5.0f));
+    else if(event.key() == KeyEvent::Key::Up)
+        _cameraObject->rotateX(Deg(-5.0f));
+    else if(event.key() == KeyEvent::Key::Left)
+        _cameraRig->rotateY(Deg(-5.0f));
+    else if(event.key() == KeyEvent::Key::Right)
+        _cameraRig->rotateY(Deg(5.0f));
+    else if(event.key() == KeyEvent::Key::C)
+    {
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExample.cpp]: Done!

File: src/dart/DartExample.cpp [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+
+    event.setAccepted();
+}
+
+void DartExample::updateGraphics() {
+    /* We refresh the graphical models at 60Hz */
+    _dartWorld->refresh();
+
+    /* For each update object */
+    for(auto& object : _dartWorld->updatedShapeObjects()){
```
[the remark this answers is not in the document]
Comment (costashatz) [reply src/dart/DartExample.cpp]: Right!

File: src/dart/DartExample.cpp (hunk 1 above)
Comment (costashatz) [reply src/dart/DartExample.cpp]: Done!

File: src/dart/DartExample.cpp (hunk 2 above)
Comment (costashatz) [reply src/dart/DartExample.cpp]: Sure!

File: src/dart/DartExample.cpp (hunk 3 above)
Comment (costashatz) [reply src/dart/DartExample.cpp]: I have only put the contents of `updateGraphics()` in the `drawEvent()` as the the `updateManipulator()` is not directly related to any graphics. Of course, we need to call it within the `drawEvent()`, but I think it'd be easier for people to follow that way.

**Comment by mosra** [comment]
Merged as 8a8ce58f240eb9d1f5ada5377db127640497fc1a. I'm happy that this is finally in! :tada:

**Pull request closed without merging**